fb_credentials.py

```diff
@@ -48,27 +49,63 @@ def get_credentials(hgrc=None, hgPrefix='', interactive=True):
             password = getpass.getpass('password: ')# Same as raw_input but hide what user types
     return username, password
 
-def FogBugz(fbConstructor, hostname, token=None, username=None, password=None, hgrc=None,
-            hgPrefix='', interactive=True, storeCredentials=False):
+def get_token(fogbugzrc=None, fogbugzPrefix=''):
+    """When credentials are not provided in the constructor, get them from fogbugzrc or prompt user
+       fogbugzrc: Path to fogbugzrc file
+       fogbugzPrefix: prefix for token. Useful if the fogbugzrc is used for multiple servers
+                 with different credentials
+    """
+    # Search whether there is an fogbugzrc file. Default: ~/.fogbugzrc
+    token = None
+    if not fogbugzrc:
+        fogbugzrc = os.path.join(os.path.expanduser("~"), '.fogbugzrc')
+    if os.path.isfile(fogbugzrc):
+        for line in open(fogbugzrc):
+            line = line.split('#')[0]
+            if fogbugzPrefix + 'token' in line:
+                res = re.search(r'token\s*=\s*(\S+)', line)
+                token = res.group(1)
+    return token
+
+def validate_token(hostname, token):
+    """
+    Validate the user token.
+
+    Returns True for a successful validation.
+    """
+    url = hostname + ('/' if hostname[-1] != '/' else '') + "api.asp?cmd=logon&token=" + token
+    try:
+        response = urllib2.urlopen(url)
+        return token in response.read()
+    except Exception as e:
+        print e
+    return False
+
+def FogBugz(fbConstructor, hostname, token=None, username=None, password=None, fogbugzrc=None,
+            fogbugzPrefix='', interactive=True, storeCredentials=False):
     """Calls the constructor specified by fbConstructor (hence, despite this being a function use
         CapWords naming convention)
 
        fbConstructor: Fogbugz constructor class. Typically fogbugz.FogBugz, fborm.FogBugzORM or
                        kiln.Kiln
        hostname: passed directly to the fbInterface
        token, username, password: input credentials
-       hgrc, hgPrefix, interactive: Passed to method get_credentials
+       fogbugzrc, fogbugzPrefix, interactive: Passed to method get_credentials
        storeCredentials: If active, create attributes token, username and password. This opens the
                           door to using it for login to other system, which is convenient, but the
                           programmer can also do what he wants with the password (which is bad).
        TODO: Support passing a list of args to fbConstructor
     """
     if token and (username or password):
-        raise TypeError("if you supply 'token' you cannot supply 'username' or 'password'")
+        raise TypeError("If you supply 'token' you cannot supply 'username' or 'password'")
     if (username and not password) or (not username and password):
         raise TypeError("You must supply both 'username' and 'password'")
+    if not token:
```

Comment:
Suppose we have provided username and password. We should use them instead of trying to log in with token. We need a test for this.